### ISSUE TITLE
toEnum{PinMode}: tag (11) is outside of enumeration's range (0,6)

### ISSUE THREAD

**Issue by hexapoder** [LeventErkok/hArduino#11]
_System.Hardware.Arduino.SamplePrograms.Blink> blink
*_\* hArduino:ERROR: toEnum{PinMode}: tag (11) is outside of enumeration's range (0,6)
**\* Make sure your Arduino is connected to /dev/cu.usbmodemFD131
**\* And StandardFirmata is running on it!

I get above failure when i am testing "blink". Firmatatest is running fine. I am a bit newbie and i couldn't debug this.  On the other side i didn't change the code.  Could there be a chance that new versions of  standartfirmata is not compatible ?  I used

OS X : 10.11.3
Firmata : 2.5.2
ghc : 7.10.2
Serialport 0.4.7
Arduino Uno

Above configuration was working with Haskino,  thats why i suspect with Firmata in which haskino was using a configurable version.

**Comment by LeventErkok** [comment]
I suspect your arduino is not quite running the firmata version that I developed with.

What do you see when you run with:

     withArduino True ...

(The example has False as the first argument; which turns off debugging. True makes it output verbose messages.)

**Comment by hexapoder** [comment]
*** hArduino:ERROR: /dev/cu.usbmodemFD131: openFd: resource busy (Resource busy)
*** Make sure your Arduino is connected to /dev/cu.usbmodemFD131
*** And StandardFirmata is running on it!

**Comment by hexapoder** [comment]
I am sorry , firmatatest was running background. Now the correct messages : 

[1:65793462644] hArduino: Accessing arduino located at: "/dev/cu.usbmodemFD131"
[2:65793465650] hArduino: Started listener thread: ThreadId 76
[3:65793466671] hArduino: Sending: SystemReset <ff>
[4:65793467377] hArduino: Sending: QueryFirmware <f0 79 f7>
[5:65793468564] hArduino: Ignoring the received response: Unimplemeneted 67 []
[6:65793468955] hArduino: Ignoring the received response: Unimplemeneted 99 []
[7:65793469322] hArduino: Ignoring the received response: Unimplemeneted 71 []
[8:65793469727] hArduino: Ignoring the received response: Unimplemeneted 103 []
[9:65793470105] hArduino: Ignoring the received response: Unimplemeneted 99 []
[10:65793470559] hArduino: Ignoring the received response: Unimplemeneted 39 []
[11:65793470933] hArduino: Ignoring the received response: Unimplemeneted 103 []
[12:65793471310] hArduino: Ignoring the received response: Unimplemeneted 99 []
[13:65793471685] hArduino: Ignoring the received response: Unimplemeneted 39 []
[14:65793472056] hArduino: Ignoring the received response: Unimplemeneted 99 []
[15:65793472426] hArduino: Ignoring the received response: Unimplemeneted 103 []
[16:65793472840] hArduino: Updating digital port Port8 values with [63, 63]
[17:65793473182] hArduino: Ignoring the received response: Unimplemeneted 115 []
[18:65793473531] hArduino: Ignoring the received response: Unimplemeneted 99 []
[19:65793473871] hArduino: Ignoring the received response: Unimplemeneted 71 []
[20:65793474218] hArduino: Ignoring the received response: Unimplemeneted REPORT_ANALOG_PIN IPin14 [43]
[21:65793475341] hArduino: Ignoring the received response: Unimplemeneted 99 []
[22:65793475881] hArduino: Ignoring the received response: Unimplemeneted 103 []
[23:65793476263] hArduino: Ignoring the received response: Unimplemeneted 99 []
[24:65793476604] hArduino: Ignoring the received response: Unimplemeneted 99 []
[25:65793476967] hArduino: Ignoring the received response: Unimplemeneted 99 []
[26:65793477333] hArduino: Ignoring the received response: Unimplemeneted 71 []
[27:65793477700] hArduino: Ignoring the received response: Unimplemeneted REPORT_ANALOG_PIN IPin6 [43]
[28:65793478228] hArduino: Ignoring the received response: Unimplemeneted 67 []
[29:65793478600] hArduino: Ignoring the received response: Unimplemeneted 83 []
[30:65793478966] hArduino: Ignoring the received response: Unimplemeneted 99 []
[31:65793479332] hArduino: Ignoring the received response: Unimplemeneted 67 []
[32:65793479695] hArduino: Ignoring the received response: Unimplemeneted 83 []
[33:65793480056] hArduino: Ignoring the received response: Unimplemeneted 67 []
[34:65793480551] hArduino: Ignoring the received response: Unimplemeneted 99 []
[35:65793480937] hArduino: Ignoring the received response: Unimplemeneted 83 []
[36:65793481301] hArduino: Ignoring the received response: Unimplemeneted 99 []
[37:65793481646] hArduino: Ignoring the received response: Unimplemeneted 99 []
[38:65793481982] hArduino: Ignoring the received response: Unimplemeneted 83 []
[39:65793482319] hArduino: Ignoring the received response: Unimplemeneted 99 []
[40:65793482710] hArduino: Ignoring the received response: Unimplemeneted 99 []
[41:65793483052] hArduino: Ignoring the received response: Unimplemeneted 83 []
[42:65793483387] hArduino: Ignoring the received response: Unimplemeneted 99 []
[43:65793483721] hArduino: Ignoring the received response: Unimplemeneted 71 []
[44:65793484057] hArduino: Ignoring the received response: Unimplemeneted 169 []
[45:65793484397] hArduino: Ignoring the received response: Unimplemeneted 67 []
[46:65793484807] hArduino: Ignoring the received response: Unimplemeneted 67 []
[47:65793485147] hArduino: Ignoring the received response: Unimplemeneted 83 []
[48:65793485483] hArduino: Ignoring the received response: Unimplemeneted 99 []
[49:65793485819] hArduino: Ignoring the received response: Unimplemeneted 71 []
[50:65793486155] hArduino: Ignoring the received response: Unimplemeneted 115 []
[51:65793486493] hArduino: Ignoring the received response: Unimplemeneted 71 []
[52:65793486882] hArduino: Ignoring the received response: Unimplemeneted 99 []
[53:65793487243] hArduino: Ignoring the received response: Unimplemeneted 71 []
[54:65793487600] hArduino: Ignoring the received response: Unimplemeneted 67 []
[55:65793487966] hArduino: Ignoring the received response: Unimplemeneted 67 []
[56:65793488329] hArduino: Ignoring the received response: Unimplemeneted 99 []
[57:65793488690] hArduino: Ignoring the received response: Unimplemeneted 71 []
[58:65793489133] hArduino: Ignoring the received response: Unimplemeneted 99 []
[59:65793489502] hArduino: Ignoring the received response: Unimplemeneted 99 []
[60:65793489864] hArduino: Ignoring the received response: Unimplemeneted 67 []
[61:65793490264] hArduino: Ignoring the received response: Unimplemeneted 67 []
[62:65793490671] hArduino: Ignoring the received response: Unimplemeneted 67 []
[63:65793491125] hArduino: Ignoring the received response: Unimplemeneted 67 []
[64:65793491533] hArduino: Ignoring the received response: Unimplemeneted 67 []
[65:65793491878] hArduino: Ignoring the received response: Unimplemeneted 67 []
[66:65793492222] hArduino: Ignoring the received response: Unimplemeneted 103 []
[67:65793492597] hArduino: Ignoring the received response: Unimplemeneted 99 []
[68:65793492965] hArduino: Ignoring the received response: Unimplemeneted 99 []
[69:65793493333] hArduino: Ignoring the received response: Unimplemeneted 67 []
[70:65793493804] hArduino: Ignoring the received response: Unimplemeneted 99 []
[71:65793494179] hArduino: Ignoring the received response: Unimplemeneted 83 []
[72:65793494579] hArduino: Ignoring the received response: Unimplemeneted 71 []
[73:65793494944] hArduino: Ignoring the received response: Unimplemeneted 99 []
[74:65793495289] hArduino: Ignoring the received response: Unimplemeneted 67 []
[75:65793495636] hArduino: Ignoring the received response: Unimplemeneted 71 []
[76:65793496051] hArduino: Ignoring the received response: Unimplemeneted 99 []
[77:65793496401] hArduino: Ignoring the received response: Unimplemeneted 71 []
[78:65793496855] hArduino: Ignoring the received response: Unimplemeneted 67 []
[79:65793497311] hArduino: Ignoring the received response: Unimplemeneted 71 []
[80:65793497695] hArduino: Ignoring the received response: Unimplemeneted 169 []
[81:65793498133] hArduino: Ignoring the received response: Unimplemeneted 99 []
[82:65793498634] hArduino: Ignoring the received response: Unimplemeneted 67 []
[83:65793499083] hArduino: Ignoring the received response: Unimplemeneted 83 []
[84:65793499454] hArduino: Ignoring the received response: Unimplemeneted 99 []
[85:65793499912] hArduino: Ignoring the received response: Unimplemeneted 67 []
[86:65793500385] hArduino: Ignoring the received response: Unimplemeneted 115 []
[87:65793500746] hArduino: Ignoring the received response: Unimplemeneted 99 []
[88:65793501290] hArduino: Ignoring the received response: Unimplemeneted 99 []
[89:65793501740] hArduino: Ignoring the received response: Unimplemeneted 83 []
[90:65793502170] hArduino: Ignoring the received response: Unimplemeneted 99 []
[91:65793502597] hArduino: Ignoring the received response: Unimplemeneted 99 []
[92:65793502983] hArduino: Ignoring the received response: Unimplemeneted 115 []
[93:65793503375] hArduino: Ignoring the received response: Unimplemeneted 99 []
[94:65793503906] hArduino: Ignoring the received response: Unimplemeneted 71 []
[95:65793504306] hArduino: Ignoring the received response: Unimplemeneted 115 []
[96:65793504655] hArduino: Ignoring the received response: Unimplemeneted 67 []
[97:65793505003] hArduino: Ignoring the received response: Unimplemeneted 67 []
[98:65793505354] hArduino: Ignoring the received response: Unimplemeneted 115 []
[99:65793505706] hArduino: Ignoring the received response: Unimplemeneted 99 []
[100:65793506109] hArduino: Ignoring the received response: Unimplemeneted 103 []
[101:65793506472] hArduino: Ignoring the received response: Unimplemeneted REPORT_ANALOG_PIN IPin14 [47]
[102:65793506937] hArduino: Ignoring the received response: Unimplemeneted 67 []
[103:65793507290] hArduino: Ignoring the received response: Unimplemeneted 67 []
[104:65793507641] hArduino: Ignoring the received response: Unimplemeneted 71 []
[105:65793508602] hArduino: Ignoring the received response: Unimplemeneted 99 []
[106:65793509088] hArduino: Ignoring the received response: Unimplemeneted 71 []
[107:65793509449] hArduino: Ignoring the received response: Unimplemeneted 99 []
[108:65793509792] hArduino: Ignoring the received response: Unimplemeneted 67 []
[109:65793510136] hArduino: Ignoring the received response: Unimplemeneted 67 []
[110:65793510563] hArduino: Ignoring the received response: Unimplemeneted 103 []
[111:65793510957] hArduino: Ignoring the received response: Unimplemeneted 99 []
[112:65793511383] hArduino: Ignoring the received response: Unimplemeneted 71 []
[113:65793511733] hArduino: Ignoring the received response: Unimplemeneted 71 []
[114:65793512080] hArduino: Ignoring the received response: Unimplemeneted 99 []
[115:65793512422] hArduino: Ignoring the received response: Unimplemeneted 103 []
[116:65793512768] hArduino: Ignoring the received response: Unimplemeneted 67 []
[117:65793513110] hArduino: Ignoring the received response: Unimplemeneted 99 []
[118:65793513505] hArduino: Ignoring the received response: Unimplemeneted 83 []
[119:65793513856] hArduino: Ignoring the received response: Unimplemeneted 103 []
[120:65793514203] hArduino: Ignoring the received response: Unimplemeneted 99 []
[121:65793514548] hArduino: Ignoring the received response: Unimplemeneted 71 []
[122:65793514888] hArduino: Ignoring the received response: Unimplemeneted 71 []
[123:65793515234] hArduino: Ignoring the received response: Unimplemeneted 99 []
[124:65793515632] hArduino: Ignoring the received response: Unimplemeneted 71 []
[125:65793515982] hArduino: Ignoring the received response: Unimplemeneted 103 []
[126:65793516331] hArduino: Ignoring the received response: Unimplemeneted 99 []
[127:65793516675] hArduino: Ignoring the received response: Unimplemeneted 39 []
[128:65793517021] hArduino: Ignoring the received response: Unimplemeneted 103 []
[129:65793517369] hArduino: Ignoring the received response: Unimplemeneted 99 []
[130:65793517765] hArduino: Ignoring the received response: Unimplemeneted 39 []
[131:65793518116] hArduino: Ignoring the received response: Unimplemeneted 99 []
[132:65793518479] hArduino: Ignoring the received response: Unimplemeneted 103 []
[133:65793518832] hArduino: Updating digital port Port8 values with [63, 63]
[134:65793519171] hArduino: Ignoring the received response: Unimplemeneted 115 []
[135:65793519525] hArduino: Ignoring the received response: Unimplemeneted 99 []
[136:65793519946] hArduino: Ignoring the received response: Unimplemeneted 71 []
[137:65797119859] hArduino: Ignoring the received response: Unimplemeneted REPORT_ANALOG_PIN IPin14 [f9]
[138:65797120371] hArduino: Ignoring the received response: Unimplemeneted 2 []
[139:65797120724] hArduino: Ignoring the received response: Unimplemeneted 5 []
[140:65797128125] hArduino: Received Firmware v2.5 (StandardFirmata.ino)
[141:65797128495] hArduino: Sending: CapabilityQuery <f0 6b f7>
[142:65797165095] hArduino: Received Capabilities:
IPin0: 
IPin1: 
IPin2: INPUT <interactive>: toEnum{PinMode}: tag (11) is outside of enumeration's range (0,6)
*** hArduino:ERROR: thread blocked indefinitely in an MVar operation
*** Make sure your Arduino is connected to /dev/cu.usbmodemFD131
*** And StandardFirmata is running on it!

**Comment by LeventErkok** [comment]
That suggests you're not actually running StandardFirmata on the board. Did you actually do that step?

    Install the Arduino environment: http://arduino.cc/en/Main/Software.

    Start the Arduino IDE. Using the 'File->Examples->Firmata' menu, upload "StandardFirmata" to your board.

If you uploaded a different Firmata (there's a few different kinds) it wouldn't work.

If you did indeed upload StandardFirmata; I'll need to check when I get home what sort of response I get from that query and compare in detail.

**Comment by hexapoder** [comment]
Well i am sure that i used the arduino ide to upload the standardFirmata in which you refer in examples. Then i tested with firmatatest.

**Comment by LeventErkok** [comment]
It's quite possible `StandardFirmata` changed since I worked on this; and thus my version doesn't work anymore. Bummer.

Let me see if I can upgrade my StandardFirmata and replicate the issue. Or; you can try to find StandardFirmata from a few years back and give it a shot. It appears I last tested this February 2014; so it wouldn't  surprise me if there were changes to it that hArduino doesn't understand.

**Comment by hexapoder** [comment]
Actually that , i was looking for in your readme , the version you developed, before opening the issue to verify. Otherwise it is my newbie mistake.

**Comment by LeventErkok** [comment]
The culprit seems to be this line:

    https://github.com/firmata/arduino/blob/master/Firmata.h#L106

They added a few new modes, and hArduino isn't aware of them; and thus is choking. This should be a relatively easy thing to fix; by mapping anything that it doesn't know to an UNKNOWN mode instead of choking.

I'll see if I can patch it up in the next few days. Thanks for bringing this to my attention.

**Comment by hexapoder** [comment]
Nice! . By the way i just picked up an older version to test . Yes. your code works fine with 2.3.6 perfectly.
I will be waiting for the fix or I will be working with that version. 

By the way , i would like to thank you for this great  piece of software that i will spend a lot of my time with fun!

**Comment by LeventErkok** [comment]
Fantastic! A new release should come sometime over the weekend.

**Comment by LeventErkok** [comment]
@hexapoder I just pushed in a fix to make the capability processing code more robust. Can you download from github and give it a test. (I tested with the blink example; but would be good to make sure it works with something a bit more complicated as well.)

Once you give me the go-ahead; I'll make a public Hackage release.

**Comment by hexapoder** [comment]
I tested  with firmata 2.5.2 (latest on library)  with some group of leds and it works for me . Sorry i don't have ready material to make more tests. But running your examples (i.e. servo ,LCD ) without equipments attached gives no error messages. Thanks.

**Comment by hexapoder** [comment]
I tried with  Linux.  I am able to install  hArduino-0.9 via "cabal install harduino" from hackage without issues. 

But , i get below failure message when "cabal install -v "  new version.  


[ 3 of 24] Compiling System.Hardware.Arduino.Data ( System/Hardware/Arduino/Data.hs, dist/build/System/Hardware/Arduino/Data.o )

System/Hardware/Arduino/Data.hs:223:38:
    Not in scope: type constructor or class `Applicative'
Failed to install hArduino-1.0
cabal: Error: some packages failed to install:
hArduino-1.0 failed during the building phase. The exception was:
ExitFailure 1

**Comment by LeventErkok** [comment]
@hexapoder This sounds like an issue with your GHC version. Are you using 7.10.3?

Also, if you could tell me your full name; I can add it to the "Thanks" list..

**Comment by hexapoder** [comment]
Worse  I seem to have  GHC ver 7.6.3 . Sorry. 

Name : Can Akçura

**Comment by LeventErkok** [comment]
Yeah; you probably need at least something in the 7.8 series; 7.10 recommended.

I'll make a release later today after some extra testing. Thanks for the report.

**Comment by LeventErkok** [comment]
@hexapoder I did some extra testing, and it appears hArduino does **NOT** fully work with the latest version of StandardFirmata. They seem to have moved to a slightly different model of updating digital-pin-values; and hArduino gets confused with the new interface.

Unfortunately I have somewhat lost interest in this project, and no longer have the manpower to keep it upto date with StandardFirmata. However, I will make a new release, pointing out that people interested should get an older copy of `StandardFirmata.ino` that I made a copy available on github.

This is a bummer, but not totally unexpected. Things move on, and it's hard to keep upto date with newer versions of everything else. If you take a shot and make hArduino work with the latest StandardFirmata, I'd be happy to accept patches!
